### PR TITLE
Community Destroyer (first CEA PR!)

### PR DESCRIPTION
**Pull request by hibukki**
# TL;DR: What does this do?
Allows CEA to decide that all community posts only get 1/2 as much score, **for purposes of sorting posts on the front page**.
## It's a bit more generic
You can decide for any tag, to multiply it's score by any amount
## Interface for changing the scores

```
const tagScoreMultipliers = [
    {
      tagId: "ZCihBFp5P64JCvQY6", // Community
      scoreMultiplier: 0.5
    },
  ]
```

### It's hardcoded
It would be better in the DB (or at least an environment variable) (or at least another file)
### It's specific to the tagId that exists in the DB
So you'll probably need a different hardcoded tagID for dev/production, which will be annoying, sorry

# Tests
## Automatic tests (jest)
None of that, sorry. It seemed hard to figure out how to test this (but if someone will give me a hand, I'll happily add a test)
### Before making any code change
Screen Shot 2022-05-24 at 22 44 03

## With the change: 1000x points for biosecurity

The code (ran in dev env) :

```
{
 tagId: "aELNHEKtcZtMwEkdK", // Biosecurity
 scoreMultiplier: 1000
}
```

Note a new post is almost on top, and it's tagged biosecurity:
Screen Shot 2022-05-24 at 22 44 29

## With the change: 1000x LESS points for biosecurity
Note the top post from before (125 karma, tagged biosecurity) is not in the feed anymore:
Screen Shot 2022-05-24 at 22 46 25

# Whitespace fixes
Sorry, if you hate it - it can be removed

# Authors
This was pair programmed with [Soof Golan](https://github.com/soof-golan)

# Before merging
## Consider merging this:

```
const tagScoreMultipliers = [
  ]
```

## Instead of this:

```
const tagScoreMultipliers = [
    {
      tagId: "ZCihBFp5P64JCvQY6", // Community
      scoreMultiplier: 0.5
    },
  ]
```

## Why?
Because maybe the community tag ID will be different in production,
and maybe you'd like the commit that configures the `tagScoreMultipliers` to be separate anyway

# Open to pushback and changes of course!
This felt more like "figure out what's going on" than "tidy development", perhaps we missed something

### CODE REVIEW

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 1]
```diff
@@ -260,29 +260,65 @@ function filterSettingsToParams(filterSettings: FilterSettings): any {
...
+      scoreMultiplier: 0.5
+    },
+  ]
   
   const tagsSoftFiltered = _.filter(filterSettings.tags, t => (t.filterMode!=="Hidden" && t.filterMode!=="Required" && t.filterMode!=="Default" && t.filterMode!==0));
   let scoreExpr: any = null;
   if (tagsSoftFiltered.length > 0 || frontpageSoftFilter.length > 0) {
     scoreExpr = {
       syntheticFields: {
-        score: {$divide:[
```
Comment (hibukki) [inline packages/lesswrong/lib/collections/posts/views.ts:269]: The changes here are only whitespace and indentation, for example `{$divide:[` became:

```
{
  $divide: [
```

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 2]
```diff
@@ -260,29 +260,65 @@ function filterSettingsToParams(filterSettings: FilterSettings): any {
...
+                              0
+                            ]
+                          }
+                        ]
+                      })),
+                      ...defaultScoreModifiers(),
+                      ...frontpageSoftFilter,
+                    ]
+                },
+                ...tagScoreMultipliers.map(t => ({
```
Comment (hibukki) [inline packages/lesswrong/lib/collections/posts/views.ts:310]: This is new code (the entire "map")

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 3]
```diff
@@ -260,29 +260,65 @@ function filterSettingsToParams(filterSettings: FilterSettings): any {
...
     scoreExpr = {
       syntheticFields: {
-        score: {$divide:[
-          {$add:[
-            "$baseScore",
-            ...tagsSoftFiltered.map(t => ({
+        score: {
+          $divide: [
+            {
               $multiply: [
```
Comment (hibukki) [inline packages/lesswrong/lib/collections/posts/views.ts:290]: "multiply" was added to multiply the scores.
Also possible: to use the "$divide" that already exists, but it might be a bit less readable.
[no idea if adding an extra function here causes performance problems in mongo, or is bad practice for some other reason]

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 4]
```diff
@@ -260,29 +260,65 @@ function filterSettingsToParams(filterSettings: FilterSettings): any {
...
+                },
+                ...tagScoreMultipliers.map(t => ({
+                  $cond: [
+                    '$tagRelevance.' + t.tagId,
+                    t.scoreMultiplier,
+                    1
+                  ]
+                })),
               ]
-            })),
```
Comment (hibukki) [inline packages/lesswrong/lib/collections/posts/views.ts:280]: The changes here are only whitespace

**Comment by jpaddison3** [comment]
Hi. I'm so sorry this wasn't communicated better, but could you submit this PR against https://github.com/ForumMagnum/ForumMagnum ? This is an old and unused fork.

**Comment by Xodarap** [comment]
Thanks for doing this! I think a database setting would be better, like you suggest – could you switch to doing that? You can search our code for references to `DatabaseServerSetting` for examples. cMy guess is that the setting should be an array of values of the form `{tagId: , multiplier: }`

These settings are configured in the databasemetadata collection on the database. I think it would make sense to put it in serverSettings.
